File: September Challenge/Week 1/Java/Day_7/WordPattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        stringstream s(str);
        map<char , string> mp;
        map<string , char> mp2;
        int i=0;
        string word;
        while(s>>word){
            if(mp.find(pattern[i])==mp.end()){
                mp[pattern[i]]=word;
            }else{
                if(mp[pattern[i]]!=word){
                    return false;
                }
            }
            
            
            if(mp2.find(word)==mp2.end()){
                mp2[word]=pattern[i];
            }else{
                if(mp2[word]!=pattern[i]){
                    return false;
                }
            }
            
            if(i==pattern.size()){
                return false;
            }
            
            i++;
        }
        if(i<pattern.size()){
            return false;
        }
        return true;
        
    }
};

File: September Challenge/Week 1/Java/Day_7/WordPattern.java
// Problem : https://leetcode.com/problems/word-pattern/
// Solution:

class Solution {
    public boolean wordPattern(String pattern, String str) {
        if(str==null || pattern == null || pattern.length()==0 || str.length()==0) return false;
        String[] s = str.split(" ");
        if(s.length!=pattern.length())  return false;
        HashMap<Character,String> map = new HashMap<>();
        
        for(int i=0;i<s.length;i++ ){
            char ch = pattern.charAt(i);
            if(!map.containsKey(ch)){
                if(!map.containsValue(s[i]))
                    map.put(ch,s[i]);
                else
                    return false;
            }else{
                if(!map.get(ch).equals(s[i])){
                    return false;
                }
            }
        }
        return true;
    }
}
